### PR TITLE
net: download_client: fix errors in retry mechanism

### PR DESCRIPTION
**Pull request by jbrzozoski**
Depending on when the socket error happened, a partial buffer was
sometimes left behind causing the retry to fail immediately.
If the failure happens when there is a partial fragment in the buffer,
hand it to the application first if necessary, then clear the buffer
before restarting.

Signed-off-by: Justin Brzozoski <justin.brzozoski@signal-fire.com>

### CODE REVIEW

**Comment by jbrzozoski** [comment]
Would it make more sense to discard the partial fragment and rewind the progress meter instead of handing a partial fragment event to the consumer?

**Comment by rlubos** [comment]
> Would it make more sense to discard the partial fragment and rewind the progress meter instead of handing a partial fragment event to the consumer?

The download_client API does not seem to make any assumptions of the fragment size notified to the upper layer, so I think it should be fine to report a partial fragment. 

@lemrey What's your say on that, are you aware of any restrictions in modules that use this library?

**Comment by jbrzozoski** [comment]
Did this miss the cutoff for v1.1 or might it still be put on that branch?

**Comment by jbrzozoski** [comment]
I'm getting intermittent VERIFY FAIL errors right around retries w/ this feature as-is.  

What's the minimum write size for flash on the nRF9160?  WORD or DWORD or ???  

I suspect the partial fragment being handed up is not ending on the proper alignment and writing extra bytes to non-0xFF that then make the resumption unable to write the proper value...

If this is the issue, we could fix it by discarding the unaligned bytes at the end of the partial fragment and rewinding progress by the same amount, or we could just discard the whole fragment.

Any opinions or wisdom from your team?

**Comment by jbrzozoski** [comment]
Or perhaps it's as simple as the fact that the none of the fota_download subsystem will ensure the flash is actually erased unless you set `CONFIG_IMG_ERASE_PROGRESSIVELY=y`?  (That can't be true, I must be missing something...)

**Comment by jbrzozoski** [comment]
> Or perhaps it's as simple as the fact that the none of the fota_download subsystem will ensure the flash is actually erased unless you set `CONFIG_IMG_ERASE_PROGRESSIVELY=y`? (That can't be true, I must be missing something...)

Okay.  That is what it was.  I think it might be wise to add an erase to flash_img_init() for the whole region in the case that `CONFIG_IMG_ERASE_PROGRESSIVELY=n`...

**Comment by lemrey** [comment]
Hi, thanks for submitting this patch; I think I understand the issue, and it is important to zero `has_header` if we hit a socket error, otherwise we will fail to parse the next bytes. We should send whatever bytes we have in the buffer to the application, especially because fragment size might be several KB.

I have piled up a commit that is a slight refactoring of your proposed patch, that tries to simplify it a bit; plus, a very tiny doc commit.

If it looks good to you, could you please squash my commit into yours (not the doc one), add my signed off message, and reword the title to add the `lib` part in it e.g. `net: lib: download_client`

File: subsys/net/lib/download_client/src/download_client.c [hunk 1]
```diff
@@ -501,18 +516,18 @@ void download_thread(void *client, void *a, void *b)
...
 		if (dl->connection_close) {
 			dl->connection_close = false;
 			reconnect(dl);
 		}
 
+		/* Request next fragment */
 		/* Send a GET request for the next bytes */
 send_again:
+		dl->offset = 0;
+		dl->has_header = false;
```
Comment (lemrey) [inline subsys/net/lib/download_client/src/download_client.c]: I moved these below `send_again` to simplify.

**Comment by jbrzozoski** [comment]
Your adjustments all look good to me.

**Comment by rlubos** [comment]
> Did this miss the cutoff for v1.1 or might it still be put on that branch?

I think we should try to integrate this PR into 1.1.0 as it fixes a (quite serious) bug.

**Pull request merged**